File: JavaLibrary/src/main/java/com/flat/simplex/parser/logic/block/BlockElse.java
package com.flat.simplex.parser.logic.block;

import com.flat.simplex.lexer.Key;
import com.flat.simplex.lexer.Token;
import com.flat.simplex.parser.Parser;
import com.flat.simplex.parser.logic.Block;
import com.flat.simplex.parser.logic.error.Error;

import java.util.ArrayList;

public class BlockElse extends Block {

    private Token tokenContent;
    private Token tokenContentEnd;
    private boolean commandBlock;

    private ArrayList<Block> blocks;

    public BlockElse(Block parent, Token start, Token end) {
        super(parent, start);

        Token token = start;
        Token lToken = start;
        int state = 0;
        while (token != end && token != null) {
            if (state == 0 && token.getKey() == Key.Else) {
                state = 1;
            } else if (state == 1 && token.getKey() == Key.Brace) {
                state = 2;
                tokenContent = token;
                commandBlock = true;
            } else if (state == 1) {
                state = 2;
                tokenContent = token;
                tokenContentEnd = end;
                break;
            } else {
                error(token, Error.unexpectedToken);
            }
            lToken = token;
            token = token.getNext();
        }
        if (state < 2) {
            error(lToken, Error.unexpectedEndOfTokens);
        }
    }

    @Override
    public void read() {
        if (tokenContent != null) {
            if (commandBlock) {
                blocks = new Parser(this).parse(tokenContent.getChild(), tokenContent.getLastChild());
                if (tokenContent.getLastChild() == null) {
                    error(tokenContent, Error.missingCloser);
                }
            } else {
                blocks = new Parser(this).parse(tokenContent, tokenContentEnd);
            }
        }
    }

    @Override
    public void setPreviousBlock(Block blockPrevious) {
        if (!(blockPrevious instanceof BlockIf) && !(blockPrevious instanceof BlockElseIf)) {
            error(getToken(), Error.elseOutOfPlace);
        }
    }

    public Token getTokenContent() {
        return tokenContent;
    }

    public Token getTokenContentEnd() {
        return tokenContentEnd;
    }

    public boolean isCommandBlock() {
        return commandBlock;
    }

    public ArrayList<Block> getBlocks() {
        return blocks;
    }
}

File: SimplexApp/simplex/parser/logic/block/BlockElse.js
var Key, Parser, BlockIf, BlockElseIf, Error;
const Block = require("simplex/parser/logic/Block.js");

class BlockElse extends Block {
    constructor(parent, start, end) {
        super(parent, start);

        this.tokenContent = null;
        this.tokenContentEnd = null;
        this.commandBlock = false;
        this.blocks = null;

        let token = start;
        let lToken = start;
        let state = 0;
        while (token !== end && token !== null) {
            if (state === 0 && token.getKey() === Key.Else) {
                state = 1;
            } else if (state === 1 && token.getKey() === Key.Brace) {
                state = 2;
                this.tokenContent = token;
                this.commandBlock = true;
            } else if (state === 1) {
                state = 2;
                this.tokenContent = token;
                this.tokenContentEnd = end;
                break;
            } else {
                this.error(token, Error.unexpectedToken);
            }
            lToken = token;
            token = token.getNext();
        }

        if (state < 2) {
            this.error(lToken, Error.unexpectedEndOfTokens);
        }
    }

    read() {
        if (this.tokenContent !== null) {
            if (this.commandBlock) {
                this.blocks = new Parser(this).parse(this.tokenContent.getChild(), this.tokenContent.getLastChild());
                if (this.tokenContent.getLastChild() === null) {
                    this.error(this.tokenContent, Error.missingCloser);
                }
            } else {
                this.blocks = new Parser(this).parse(this.tokenContent, this.tokenContentEnd);
            }
        }
    }

    setPreviousBlock(blockPrevious) {
        if (!(blockPrevious instanceof BlockIf) && !(blockPrevious instanceof BlockElseIf)) {
            this.error(this.getToken(), Error.elseOutOfPlace);
        }
    }

    getTokenContent() {
        return this.tokenContent;
    }

    getTokenContentEnd() {
        return this.tokenContentEnd;
    }

    isCommandBlock() {
        return this.commandBlock;
    }

    getBlocks() {
        return this.blocks;
    }
}

module.exports = BlockElse;
Key = require("simplex/lexer/Key.js");
Parser = require("simplex/parser/Parser.js");
BlockIf = require("simplex/parser/logic/block/BlockIf.js");
BlockElseIf = require("simplex/parser/logic/block/BlockElseIf.js");
Error = require("simplex/parser/logic/error/Error.js");